transaction/std/src/transaction_builder.rs

```diff
@@ -321,6 +334,64 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
         result
     }
 
+    /// Add an output to the reserved subaddress for gift codes
+    ///
+    /// The gift code subaddress is meant for reserving TxOuts for usage
+    /// at a later time. This method creates outputs to that address in
+    /// a way that Fog can track by creating a Fog hint for the primary
+    /// account. This allows Fog users who send TxOuts to this address to
+    /// track reserved TxOuts and if they desire, let other Fog users find
+    /// these TxOuts and spend them at a later time. This enables
+    /// functionality like sending "gift codes" to individuals who may not
+    /// have a MobileCoin account and "red envelopes".
+    ///
+    /// The caller should ensure that the math adds up, and that
+    /// change_value + gift_code_amount + fee = total_input_value
+    ///
+    /// # Arguments
+    /// * `amount` - The amount of the "gift code"
+    /// * `reserved_subaddreses` - A ReservedSubaddresses object which
+    /// provides all standard reserved addresses for the caller. This is
+    /// used to set the caller's primary address as the Fog hint address
+    /// and set their gift code subaddresses as the TxOut recipient.
+    /// * `rng` - RNG used to generate blinding for commitment
+    pub fn add_gift_code_output<RNG: CryptoRng + RngCore>(
+        &mut self,
+        amount: Amount,
+        reserved_subaddresses: &ReservedSubaddresses,
+        rng: &mut RNG,
+    ) -> Result<(TxOut, TxOutConfirmationNumber), TxBuilderError> {
+        // Taking self.memo_builder here means that we can call functions on &mut self,
+        // and pass them something that has captured the memo builder.
+        // Calling take() on Option<Box> is just moving a pointer.
+        let mut mb = self
+            .memo_builder
+            .take()
+            .expect("memo builder is missing, this is a logic error");
+        let block_version = self.block_version;
+        let result = self.add_output_with_fog_hint_address(
+            amount,
+            &reserved_subaddresses.gift_code_subaddress,
+            &reserved_subaddresses.primary_address,
+            |memo_ctxt| {
+                if block_version.e_memo_feature_is_supported() {
+                    Some(mb.make_memo_for_output(
```

Comment:
I think it would be slightly better if this were `make_memo_for_gift_code_output` because it would catch some more bugs. But it's also more code, I'm okay if we don't do this.

Reply:
I will do that, but let me scope it to an immediate PR, so people can weigh in on it in that context.

Reply:
Actually I think you're right that it's a lot of code. We'd have to add that as a trait method for every trait for it to be called once, so I think it might actually lend to more confusion in that configuration.